### ISSUE TITLE
Strategy, Tags & Branches

### ISSUE THREAD

**Issue by innovaker** [zmkfirmware/zmk-docker#62]
## Now
Currently ...
- in git and Docker, the tag `2.4` means the "most recent 2.4 image that has been (manually) verified".  So does "2.4.0-0.11.4".
- in Docker, the branch name - e.g. `main` or `a/pr/branch` - means "most recently built for that branch" (+ inline cache).
- in git, branch names have no meaning.

All other tags are fully qualified candidates.

## Future

Going forward, we'll probably need branches per major/minor Zephyr release.  Not as traditional maintainence branches per se, but because I think this is the only way to schedule independent nightly builds in GitHub?
See: https://discord.com/channels/719497620560543766/719544230497878116/846751289962266704

### Considerations
- Do we even need a `main` branch?
  - `zmk-docker` will always track with Zephyr releases.
  - `major.minor` (`2.4`) branches would probably suffice.
    - each one feeds into the next
    - we change the default branch when Zephyr increments its major or minor version
    - it simplifies the candidate Docker tag naming conventions
  - Every PR is likely to be made against a specific branch.
    - Are version-spanning PR's likely to occur?

- Will we need `major.minor.revision` (`2.4.1`) tags or branches?
  - tags - probably wise to do this anyway.  So we have `2`, `2.4` and `2.4.1` Docker tags.
  - branches - maybe?  Perhaps in the future?

- Should Docker tags be the same as the git tag/branches?
  - it's probably the most intuitive approach.

- Should git tags use the `v` prefix?
  - many projects use this
    - some because they're old
    - some because they copy others
    - some by choice
  - some popular projects don't!
    - e.g. rust, angular, etc.
  - it's largely associated with semantic versioning (even though it's not in the specification!)
  - some argue it's a hangover from limitations of deprecated VCS
  - there are potential benefits either way
  - we're not actually using semantic versioning for `zmk-docker`'s image releases

- The candidate tags should *probably* either be an extension of the branch tag or the Zephyr version.
  - this makes it easier to understand that they're snapshots

- We should use the word `latest` with great care.
  - the `latest` Docker tag is ambiguous and often misunderstood or misused
  - each Docker repository tends to have its own meaning for `latest`
  - if someone specifies the Docker image without a tag, Docker looks for the image tagged `latest`
  - so, should `latest` - either on its own or as a suffix - refer to ...
    - the most recently pushed image (the last build on that branch)?
    - the most recently verified image (manually tested)?

---

I've looked at a lot of popular git repositories and Docker repositories.  Each has their own approach and style.

IMHO, `zmk-docker` has quite a unique approach and purpose compared to most projects.  We should therefore use whatever meets our needs rather than copying existing examples.

My working exploration of possibilities is here:
https://docs.google.com/spreadsheets/d/1F_kNvqk6SIGD2ulBJx43i1bWw-gMD6i7Unq4z94KRXM/

**Comment by innovaker** [comment]
This is still mulling around my head, but so far ...

---

> Should git tags use the v prefix?

I'm leaning towards not using a `v` prefix, because we're not actually versioning in the strictest sense.

---

> My working exploration of possibilities is here:
> https://docs.google.com/spreadsheets/d/1F_kNvqk6SIGD2ulBJx43i1bWw-gMD6i7Unq4z94KRXM/

I was originally keen on the `2.4.x` notation, but I realised that the potential for `2.4.1.x` branches (for whatever reason?) would be weird.

`2.4-branch` seems the least controversial aside from being boring.

---

Everything else I'm still considering from different angles.

**Comment by innovaker** [comment]
> Do we even need a main branch?

Given the nature of branches, I'm leaning towards giving a `main`-less approach a trial run.

**Comment by petejohanson** [comment]
> > Do we even need a main branch?
> 
> Given the nature of branches, I'm leaning towards giving a `main`-less approach a trial run.

My only question is how we handle changes like "add gdb". Without a `main` it seems like we just select whatever the latest version branch is we are targeting? Make it the default branch in GH and call it a day?

Suppose that works. Seems odd, but might as well.

**Comment by innovaker** [comment]
> Without a main it seems like we just select whatever the latest version branch is we are targeting? Make it the default branch in GH and call it a day?

Yep.  You can also change which branch a PR is based on (the target) by editing it ...
![image](https://user-images.githubusercontent.com/66737976/120174230-c5d63000-c1fc-11eb-8b92-45887a5b351f.png)
so I think it would work.  It's just unusual.

That said, I mainly (pun intended) floated the idea because the branch name appears in the Docker tags and candidate git tags.  At first glance it seemed a little awkward ...
- `main` builds nightly: `main-##############-2.5.0-0.12.4...`
- "release" `2.5` by specifying one of the candidate/nightly builds
  - git (trigger) tag would be `main-##############-2.5.0-0.12.4...`
- create `2.5` branch based on `main`'s `HEAD` or the commit which spawned the candidate?
  - automated or manually?
  - cache is created from scratch
  - `2.5` branch builds nightly: `2.5-##############-2.5.0-0.12.4...`
- increase `main` to `zephyr-version: v2.6.0-rc2`
  - `main` builds nightly: `main-##############-2.6.0-rc2-0.12.4...`
- increase `main` to `zephyr-version: v2.6.0-rc3`
  - `main` builds nightly: `main-##############-2.6.0-rc3-0.12.4...`
- increase `main` to `zephyr-version: v2.6.0`
  - `main` builds nightly: `main-##############-2.6.0-0.12.4...`
- "release" `2.6` by specifying one of the candidate/nightly builds
  - git (trigger) tag would be `main-##############-2.6.0-0.12.4...`
- create `2.6` branch based on `main`'s `HEAD` or the commit which spawned the candidate?
  - automated or manually?
  - cache is created from scratch
  - `2.6` branch builds nightly: `2.6-##############-2.6.0-0.12.4...`

etc.

But having thought about it some more, `main` is probably fine as long as we keep the explicit `2.x.0-0.12.4` in the tags.  It's probably easier than changing the default GH branch per release - which would probably get forgotten at some point.

**Comment by innovaker** [comment]
Here's an interesting twist ... on https://github.com/innovaker/zmk-docker the `main` and `future` branches are currently more or less the same (for testing).  But only `main` seems to be building nightly.  Perhaps there's some undocumented constraint around `schedule`.

**Comment by innovaker** [comment]
>Perhaps there's some undocumented constraint around schedule.

It's documented on a different page:
https://docs.github.com/en/actions/reference/events-that-trigger-workflows#scheduled-events

TLDR: schedule only runs on the latest commit of the *default branch*.

Hmmmm, I guess that changes the ballgame slightly ... 💭

**Comment by innovaker** [comment]
To summarize, we can ... 
- only `schedule` nightly builds on the default branch
- trigger rebuilds manually (`workflow_dispatch`) on any branch
  - can be triggered via the REST API

So nightly builds on any branch are still theoretically feasible, but would need to be triggered remotely (i.e. a script on a remote `cron` job).

**Comment by innovaker** [comment]
Given the above revelation, it strikes me that the default branch (`main` or otherwise) should always be the "current" version of Zephyr rather than the "next" version.  So that the `schedule` builds always fulfil their primary purpose: to deliver the latest security updates to those that want them.

This means that "next" builds - which testers will care about - will be built via PRs (built by a fork) or perhaps secondary branches.  And we only bring them into the mainstream (by merging or changing the default branch) "just in time" when it's ready.  Whether or not that should be built into the release process is debatable.

It also means that "past" versions are unlikely to receive security updates.  Which isn't really an issue given that we use trunk-based development on `zmk` and only support "current".  If someone needs this, they can achieve it by forking and changing their default branch.

**Comment by innovaker** [comment]
The above would now look like this with a `main` branch ...

*Trunk-based strategy* (which we effectively use for `zmk`) ...
- `main` builds nightly: `main-##############-2.5.0-0.12.4...`
- create `2.6` branch/PR based on `main`'s `HEAD`
- increase `2.6` branch/PR to `zephyr-version: v2.6.0-rc2`
    - builds `2.6-##############-2.6.0-rc2-0.12.4...`
- increase `2.6` branch/PR to `zephyr-version: v2.6.0-rc3`
    - builds `2.6-##############-2.6.0-rc3-0.12.4...`
- increase `2.6` branch/PR to `zephyr-version: v2.6.0`
    - builds `2.6-##############-2.6.0-0.12.4...`
- verify candidate
- merge `2.6` branch/PR into `main`
  - builds:  `main-##############-2.6.0-0.12.4...`
- `main` builds nightly: `main-##############-2.6.0-0.12.4...`
- verify new candidate
- "release" `2.6` by specifying one of the candidate/nightly builds
  - git (trigger) tag would be `main-##############-2.6.0-0.12.4...`

... or ...

*GitHub flow* strategy ...
- `main` builds nightly: `main-##############-2.5.0-0.12.4...`
- create `2.6` branch/PR based on `main`'s `HEAD`
- increase `2.6` branch/PR to `zephyr-version: v2.6.0-rc2`
  - builds `2.6-##############-2.6.0-rc2-0.12.4...`
- increase `2.6` branch/PR to `zephyr-version: v2.6.0-rc3`
  - builds `2.6-##############-2.6.0-rc3-0.12.4...`
- increase `2.6` branch/PR to `zephyr-version: v2.6.0`
  - builds `2.6-##############-2.6.0-0.12.4...`
- verify candidate
- "release" `2.6` by specifying one of the candidate/nightly builds
  - git (trigger) tag would be `2.6-##############-2.6.0-0.12.4...`
- merge `2.6` branch/PR into `main`
  - builds:  `main-##############-2.6.0-0.12.4...`
- `main` builds nightly: `main-##############-2.6.0-0.12.4...`

Note the subtle differences between the two.

We'd probably prefer the *GitHub flow* strategy in this instance to minimize testing (risk of regression is minimal).

**Comment by petejohanson** [comment]
@innovaker Why don't we have the scheduled job be configured to dispatch workflow runs to multiple maintenance branches, so we can do nightly builds for multiple versions?

**Comment by petejohanson** [comment]
That being said, I do agree the `main`-less version seems like the route to go, either way.

**Comment by innovaker** [comment]
> @innovaker Why don't we have the scheduled job be configured to dispatch workflow runs to multiple maintenance branches, so we can do nightly builds for multiple versions?

Good idea!  Let's keep that one up our sleeve for when the need for maintainence branches arises.

**Comment by innovaker** [comment]
I think the posts above have covered most of my original consideration points, aside from `latest` which I've been ignoring because it's awkward.  @petejohanson, any other thoughts please?

**Comment by petejohanson** [comment]
> I think the posts above have covered most of my original consideration points, aside from `latest` which I've been ignoring because it's awkward.  @petejohanson, any other thoughts please?

No other thoughts, and I think we skip `latest` for the time being. That tracks with `main`-less also, there's *only* explicit versions of things

**Comment by innovaker** [comment]
A recap of the conclusions ...

---

> Do we even need a `main` branch?

We're going to try `main`-less and see how it feels.

---

> Will we need `major.minor.revision` (`2.4.1`) tags or branches?

Maybe in the future, so our approach should be flexible.

---

> Should Docker tags be the same as the git tag/branches?

Yes, so it's intuitive.

---

> Should git tags use the `v` prefix?

No, but we can probably change it if the need arises.

---

> The candidate tags should *probably* either be an extension of the branch tag or the Zephyr version.

Candidates tags will extend the branch tag.

---

> We should use the word `latest` with great care.

We're going to avoid/skip it for now.

---

So:
- The git release tags (and corresponding Docker tags) will be in the form: `2.4`
- The git branch names (and corresponding Docker tags) will be in the form: `2.4-branch`
- We can change it if we find a better approach, but there will need to be a (long) transitional period between different schemes.
